implementations/rust/ockam/ockam_command/src/state.rs

```diff
@@ -0,0 +1,535 @@
+        let identifier = identity.identifier().clone();
+        let change_history = identity.change_history().await;
+        Self {
+            vault: vault_config,
+            identifier,
+            change_history,
+        }
+    }
+
+    pub async fn get(&self, ctx: &ockam::Context) -> anyhow::Result<Identity<Vault>> {
+        let data = self.change_history.export()?;
+        let vault = self.vault.get().await?;
+        Ok(Identity::import(ctx, &data, &vault).await?)
+    }
+}
+
+impl PartialEq for IdentityConfig {
+    fn eq(&self, other: &Self) -> bool {
+        self.vault == other.vault
+            && self.identifier == other.identifier
+            && self.change_history.compare(&other.change_history)
+                == IdentityHistoryComparison::Equal
+    }
+}
+
+impl Eq for IdentityConfig {}
+
+pub struct NodesState {
+    dir: PathBuf,
+}
+
+impl NodesState {
+    fn new(cli_path: &Path) -> anyhow::Result<Self> {
+        let dir = cli_path.join("nodes");
+        std::fs::create_dir_all(&dir)?;
+        Ok(Self { dir })
+    }
+
+    pub fn create(
+        &self,
+        vault: VaultState,
+        identity: IdentityState,
+        name: &str,
+    ) -> anyhow::Result<NodeState> {
+        let path = {
+            let mut path = self.dir.clone();
+            path.push(name);
+            std::fs::create_dir_all(&path)?;
+            path
+        };
+        let state = NodeState::new(path, vault, identity);
+        std::fs::write(state.path.join("version"), &state.version)?;
+        std::fs::File::create(state.socket())?;
+        std::fs::File::create(state.stdout_log())?;
+        std::fs::File::create(state.stderr_log())?;
+        std::os::unix::fs::symlink(&state.vault.path, state.path.join("vault"))?;
+        std::os::unix::fs::symlink(&state.identity.path, state.path.join("identity"))?;
+        Ok(state)
+    }
+
+    pub fn get(
+        &self,
+        vault: VaultState,
+        identity: IdentityState,
+        name: &str,
+    ) -> anyhow::Result<NodeState> {
+        let path = {
+            let mut path = self.dir.clone();
+            path.push(name);
+            if !path.exists() {
+                return Err(anyhow::anyhow!("node `{name}` does not exist"));
+            }
+            path
+        };
+        Ok(NodeState::new(path, vault, identity))
+    }
+
+    fn vault_name(&self, name: &str) -> anyhow::Result<String> {
+        let mut path = self.dir.clone();
+        path.push(name);
+        path.push("vault");
+        let path = std::fs::canonicalize(&path)?;
+        file_stem(&path)
+    }
+
+    fn identity_name(&self, name: &str) -> anyhow::Result<String> {
+        let mut path = self.dir.clone();
+        path.push(name);
+        path.push("identity");
+        let path = std::fs::canonicalize(&path)?;
+        file_stem(&path)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct NodeState {
```

Comment:
this looks like something the background running nodes would use.   Is this definition available to them?  or they code is defined on a different crate like ockam_api or something?

Reply:
We might need to move something to the ockam_api crate. I wanted to keep this first version on a single file and on the ockam_command crate to keep it as simple as possible. Ideally, the ockam_api crate would not need to know anything about this state.rs file.